### PR TITLE
Rebuild on changes to the deployment target when compiling Apple targets

### PR DESCRIPTION
**Pull request by madsmtm**
Rebuild, both in Cargo and the incremental cache, when the `*_DEPLOYMENT_TARGET` environment variables change.

Fixes https://github.com/rust-lang/rust/issues/118204. See https://github.com/rust-lang/rust/issues/129432 for more motivation behind this change.

I am a bit unsure if I have implemented this correctly, I'm particularly uncertain whether creating a new field in `Options` the correct approach for busting the incremental cache? **It's only the codegen that needs to be re-run**, everything before that shouldn't be influenced by this, so if there's a way to make that happen, we should do that instead.

Builds upon https://github.com/rust-lang/rust/pull/130068.

r? thomcc

CC @BlackHoleFox
@rustbot blocked

### CODE REVIEW

**Comment by jieyouxu** [comment]
> I'd also like some input on how to add a test for these kinds of incremental compilation changes?

If the incremental test suite doesn't suit your needs, you could fallback to a run-make test if needed.

**Comment by madsmtm** [comment]
> > I'd also like some input on how to add a test for these kinds of incremental compilation changes?
> 
> If the incremental test suite doesn't suit your needs, you could fallback to a run-make test if needed.

Thanks for the tip, I took a look at the incremental test suite, but ended up creating a `run-make` test instead. Note that this change is going to conflict with https://github.com/rust-lang/rust/pull/128507, since I updated the test from in there.

**Comment by jieyouxu** [comment]
> Note that this change is going to conflict with #128507, since I updated the test from in there.

I'll discuss this with @Oneirical, if this PR is updating the macos deployment test we might as well drop the port in https://github.com/rust-lang/rust/pull/128507.

File: compiler/rustc_driver_impl/src/lib.rs [hunk 1]
```diff
@@ -862,9 +862,9 @@ fn print_crate_info(
                 use rustc_target::spec::current_apple_deployment_target;
 
                 if sess.target.is_like_osx {
-                    let (major, minor) = current_apple_deployment_target(&sess.target)
-                        .expect("unknown Apple target OS");
-                    println_info!("deployment_target={}", format!("{major}.{minor}"))
+                    let (major, minor, patch) = current_apple_deployment_target(&sess.target);
+                    let patch = if patch != 0 { format!(".{patch}") } else { "".into() };
```
Comment (jieyouxu) [inline compiler/rustc_driver_impl/src/lib.rs:866]: Nit:

```suggestion
                    let patch = if patch != 0 { format!(".{patch}") } else { String::new() };
```

File: compiler/rustc_interface/src/passes.rs [hunk 2]
```diff
@@ -492,16 +491,22 @@ fn write_out_deps(tcx: TyCtxt<'_>, outputs: &OutputFilenames, out_filenames: &[P
 
             // Emit special comments with information about accessed environment variables.
             let env_depinfo = sess.psess.env_depinfo.borrow();
+
+            // We will soon sort, so the initial order does not matter.
+            #[allow(rustc::potential_query_instability)]
```
Comment (jieyouxu) [inline compiler/rustc_interface/src/passes.rs:496]: Question [REV1]: this looks a little bit sus. We will sort soon, but the sort below is `sort_unstable`?

File: compiler/rustc_interface/src/passes.rs [hunk 3]
```diff
@@ -492,16 +491,22 @@ fn write_out_deps(tcx: TyCtxt<'_>, outputs: &OutputFilenames, out_filenames: &[P
...
+            let mut env_depinfo: Vec<_> = env_depinfo
+                .iter()
+                .map(|(k, v)| (escape_dep_env(k.as_str()), v.map(|v| escape_dep_env(v.as_str()))))
+                .chain(tcx.sess.target.is_like_osx.then(|| {
+                    // On Apple targets, we also depend on the deployment target environment variable.
+                    let name = rustc_target::spec::apple_deployment_target_env(&tcx.sess.target.os);
+                    (name.into(), std::env::var(name).ok().map(|var| escape_dep_env(&var)))
+                }))
+                .collect();
+            env_depinfo.sort_unstable();
```
Comment (jieyouxu) [inline compiler/rustc_interface/src/passes.rs:506]: Question [REV1]: `sort_unstable` here. Can this ever cause reproducibility problems? Or rather, is there any particular reason this needs to be `sort_unstable` instead of a stable sort?

File: compiler/rustc_target/src/spec/base/apple/mod.rs [hunk 4]
```diff
@@ -97,53 +98,95 @@ impl TargetAbi {
...
+    // > The linker accepts universal (multiple-architecture) input files,
+    // > but always creates a "thin" (single-architecture), standard Mach-O
+    // > output file. The architecture for the output file is specified using
+    // > the -arch option.
+    //
+    // The linker has heuristics to determine the desired architecture, but to
+    // be safe, and to avoid a warning, we set the architecture explicitly.
+    //
+    // Supported architecture names can be found in the source:
+    // https://github.com/apple-oss-distributions/ld64/blob/ld64-951.9/src/abstraction/MachOFileAbstraction.hpp#L578-L648
```
Comment (jieyouxu) [inline compiler/rustc_target/src/spec/base/apple/mod.rs:112]: Remark: thank you for the link, this is very helpful 👍

File: tests/run-make/apple-deployment-target/rmake.rs
Comment (jieyouxu) [inline tests/run-make/apple-deployment-target/rmake.rs:1]: Remark: these comments are very helpful, thanks!

File: tests/run-make/apple-deployment-target/rmake.rs [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+    // Test that version makes it to the object file.
+    {
+        let mut rustc = rustc();
+        rustc.target(target());
+        rustc.crate_type("lib");
+        rustc.emit("obj");
+        rustc.input("foo.rs");
+        rustc.output("foo.o");
+
+        rustc.env(env_var, example_version).run();
```
Comment (jieyouxu) [inline tests/run-make/apple-deployment-target/rmake.rs:52]: Remark: not necessarily a problem, but note that all these rustc invocations will dump their output artifacts in the same directory as following invocations. If you want to make sure their output artifacts don't interfere with each other, you can use `run_in_tmpdir`, which will "quarantine" whatever invocations you run within by changing the working directory and output directory of the rustc invocations.

Comment (madsmtm) [reply tests/run-make/apple-deployment-target/rmake.rs:52]: Good point, I've changed the test to use `run_in_tmpdir`

File: compiler/rustc_interface/src/passes.rs (hunk 3 above)
Comment (madsmtm) [reply compiler/rustc_interface/src/passes.rs:506]: No idea, this is mostly existing code, I just moved it

Comment (madsmtm) [reply compiler/rustc_interface/src/passes.rs:506]: Thinking about it, I doubt that the ordering matters, this is environment variable names and values that are passed to Cargo, which in turn uses them all at the same time to determine whether to rebuild.

**Comment by petrochenkov** [comment]
Blocked on https://github.com/rust-lang/rust/pull/129341.
@rustbot blocked

**Comment by petrochenkov** [comment]
>Rebuild, both in Cargo and the incremental cache

I think the correct way to do this would be to integrate env var reading into rustc's query system.
We can make an `eval_always` query that reads the variable, then it will automatically poison all its dependent queries if the env var changes between compilations.
This query would then be used for reading *all* environment variables, not just `*_DEPLOYMENT_TARGET`.

The query can also put the read variable into `sess.psess.env_depinfo`, then it will be automatically tracked for cargo as well.

It would also be better to split the Apple-specific parts, including tests, from rustc infrastructure changes, put them into a different PR and land it separately.

**Comment by petrochenkov** [comment]
There are some issues related to target specs though.

First, the query system is not available in `rustc_target` when targets are loaded.
So we may need to re-query the relevant env vars used from there later.

Second, with `--check-cfg` we load *all* the targets, including the Apple ones, and depend on the Apple env vars.
`--check-cfg` is enabled by default now, so we technically depend on the Apple-related env vars on all hosts and targets, in most scenarios, if we consider just constructing a `Target(Option)` value a dependency.

Which fields of the target specs actually depend on the env vars?
Maybe the logic related to those fields can be excluded from target specs and moved to some other place (link.rs or where else they are necessary)?
Then the dependencies will be introduced only when strictly necessary.

**Comment by petrochenkov** [comment]
Linker arguments depending on env vars can certainly be added in link.rs, instead of target specs, likely in `fn add_order_independent_options`.

The `llvm_target` field can be versionless in target specs, the version can be added only when strictly necessary, e.g. before passing to linker or maybe to LLVM backend as well.

(Maybe I missed some other fields, didn't look very carefully.)

**Comment by madsmtm** [comment]
> It would also be better to split the Apple-specific parts, including tests, from rustc infrastructure changes, put them into a different PR and land it separately.

Good point, I've started with the tests in https://github.com/rust-lang/rust/pull/130068, will get back to your other comments.

**Comment by madsmtm** [comment]
> Linker arguments depending on env vars can certainly be added in link.rs, instead of target specs, likely in `fn add_order_independent_options`.

Yes, that's also likely what I will end up doing anyhow to implement proper `SDKROOT` reading.

> The `llvm_target` field can be versionless in target specs, the version can be added only when strictly necessary, e.g. before passing to linker or maybe to LLVM backend as well.

I... never actually really considered that, might be doable? And would be a lot cleaner, I really dislike that `rustc_target` is impure in this fashion.

> (Maybe I missed some other fields, didn't look very carefully.)

I don't think you missed any, but I think the question should rather be "which fields could we possible _want_ to depend on the deployment target in the future". That is, I _would_ like to change `rustc_target` so that it doesn't depend on these, but it has to be forwards compatible, so that future people (likely myself) that wants to configure a target specially depending on the deployment target, is empowered to do so.

I'm going through all the current `Target[Options]`, and considering whether they might want to change depending on deployment target:

<details><summary>Details</summary>
<p>

```
// Target
llvm_target -> yes, as discussed
metadata -> no
pointer_width -> no
arch -> possibly, but probably no
data_layout -> possibly

// TargetOptions
is_builtin -> no
endian -> no
c_int_width -> no
os -> no
env -> no
abi -> no
vendor -> no
linker -> no, though might want to depend on the SDK version
linker_flavor -> no
linker_flavor_json -> no
lld_flavor_json -> no
linker_is_gnu_json -> no
*link* -> yes, as discussed, but only desired to load when actually linking
asm_args -> yes
cpu -> yes, definitely
features -> yes, definitely
direct_access_external_data -> no
dynamic_linking -> possibly, watchOS might start supporting it
dll_tls_export -> no
only_cdylib -> no
executables -> no
relocation_model -> unlikely
code_model -> no
tls_model -> no
disable_redzone -> no
frame_pointer -> possibly?
function_sections -> no
dll_prefix -> no
dll_suffix -> no
exe_suffix -> no
staticlib_prefix -> no
staticlib_suffix -> no
families -> unlikely
abi_return_struct_as_int -> no
is_like_* -> no
default_dwarf_version -> yes
allows_weak_linkage -> no
has_rpath -> no
no_default_libraries -> no
position_independent_executables -> no
static_position_independent_executables -> no
plt_by_default -> no
relro_level -> no
archive_format -> no
allow_asm -> no
main_needs_argc_argv -> no
has_thread_local -> would've been useful in the past
obj_is_bitcode -> no
forces_embed_bitcode -> would've been useful in the past
bitcode_llvm_cmdline -> no
min_atomic_width -> no
max_atomic_width -> no
atomic_cas -> no
panic_strategy -> no
crt_static_allows_dylibs -> no
crt_static_default -> no
crt_static_respected -> no
stack_probes -> possibly
min_global_align -> no
default_codegen_units -> no
default_codegen_backend -> no
trap_unreachable -> no
requires_lto -> no
singlethread -> no
no_builtins -> no
default_hidden_visibility -> no
emit_debug_gdb_scripts -> no
requires_uwtable -> no
default_uwtable -> no
simd_types_indirect -> no
limit_rdylib_exports -> no
override_export_symbols -> no
merge_functions -> no
mcount -> no
llvm_mcount_intrinsic -> no
llvm_abiname -> no
relax_elf_relocations -> no
llvm_args -> possibly
use_ctors_section -> no
eh_frame_header -> no
has_thumb_interworking -> no
debuginfo_kind -> no
split_debuginfo -> no
supported_split_debuginfo -> no
supported_sanitizers -> no
c_enum_min_bits -> no
generate_arange_section -> no
supports_stack_protector -> no
entry_name -> no
entry_abi -> no
supports_xray -> no
```

</p>
</details> 

In summary, most of these are really just codegen, though some of them would be interesting for the standard library to do version detection, and emit more performant implementations if the deployment target is high enough. And in those cases, we _do_ want it to be integrated properly with the query system, such that only e.g. the MIR in the standard library that depend on those would need to be changed.

A query like `deployment_target` would also be useful for something like https://github.com/rust-lang/rfcs/pull/3379, if we had `cfg!(os_version_min(...))` we'd want incremental compilation to work with that.

All of this to say that I've convinced myself that turning it into a query is a good idea.

**Comment by madsmtm** [comment]
Just to make sure I understand correctly, I would add a `query env_var` in `compiler/rustc_middle/src/query/mod.rs`? And then in another PR I would change `llvm_target` to be versionless in target specs, and make the deployment target be loaded using the query instead? Should that also be a query like `query deployment_target`, or should that just be a regular method that calls `env_var` internally?

And since I'm asking anyways, though I'll do it separately for sure, as part of https://github.com/rust-lang/rust/issues/129432 I know that I'm also gonna need the ability to mark certain system file paths as part of the incremental cache - should this be done generally as well (`query file_changed`), or is it niche enough that it should just be its own `query sdkroot`?

**Comment by madsmtm** [comment]
> Linker arguments depending on env vars can certainly be added in link.rs, instead of target specs, likely in `fn add_order_independent_options`.

Actually, how would I go about doing that, if I wanted the deployment target to be a query? It seems like the `TyCtx<'_>` no longer exists when `add_order_independent_options` is called? Although there is `ParseSess::env_depinfo`, do you mean I should use that?

**Comment by petrochenkov** [comment]
>Just to make sure I understand correctly, I would add a query env_var in compiler/rustc_middle/src/query/mod.rs?

Yes, seems about right.

>And then in another PR I would change llvm_target to be versionless in target specs, and make the deployment target be loaded using the query instead?

Yep.

>Should that also be a query like query deployment_target, or should that just be a regular method that calls env_var internally?

Just `env_var` query should be enough for a start.

>I'm also gonna need the ability to mark certain system file paths as part of the incremental cache

This is better left for the last, I don't think we are currently doing anything like that in the compiler.

>It seems like the `TyCtx<'_>` no longer exists when `add_order_independent_options` is called?

`add_order_independent_options` has the `codegen_results.crate_info` in arguments, `tcx` is available when it is constructed.

**Comment by petrochenkov** [comment]
https://github.com/rust-lang/rust/pull/130068 has landed, unblocking.
@rustbot author

**Comment by madsmtm** [comment]
> > It seems like the `TyCtx<'_>` no longer exists when `add_order_independent_options` is called?
> 
> `add_order_independent_options` has the `codegen_results.crate_info` in arguments, `tcx` is available when it is constructed.

Ah, I see now, thanks!

I will try to find some time next week to implement what we've discussed.

**Comment by Dylan-DPC** [comment]
@madsmtm any updates on this? thanks

**Comment by petrochenkov** [comment]
https://github.com/rust-lang/rust/pull/130883 has landed, unblocking.
@rustbot author

**Comment by Dylan-DPC** [comment]
@madsmtm this is no longer blocked so any updates on this?
